### PR TITLE
Assume `@extern` functions mutate their arguments

### LINKED ISSUE #244: Add a `@pure` annotation to trigger a purity check

### LINKED ISSUE #245: @extern functions should be considered impure by default
One should be able to get back the previous behavior by adding an `@pure` annotation (cf. #244).

### PR DESCRIPTION
**Pull request by romac**
Annotate such a function with `@pure` to assume otherwise.

This PR subsumes #250, and should either be merged instead of #250, or should be rebased on top of it once it gets merged.

Closes #244. Closes #245.

### CODE REVIEW

File: core/src/main/scala/stainless/extraction/imperative/Trees.scala [hunk 1]
```diff
@@ -122,6 +123,13 @@ trait Trees extends innerfuns.Trees with Definitions { self =>
    *               EXTRACTORS
    * ======================================== */
 
+  override def extractFlag(name: String, args: Seq[Any]): Flag = (name, args) match {
+    case ("pure", Seq()) => Pure
+    case ("var", Seq()) => IsVar
+    case ("mutable", Seq()) => IsMutable
```
Comment (samarion) [inline core/src/main/scala/stainless/extraction/imperative/Trees.scala]: `var` and `mutable` don't correspond to any library annotations, so they shouldn't appear in `extractFlag`.

File: core/src/main/scala/stainless/verification/DefaultTactic.scala [hunk 2]
```diff
@@ -90,6 +90,8 @@ trait DefaultTactic extends Tactic {
   }
 
   def generateCorrectnessConditions(id: Identifier): Seq[VC] = {
+    val fd = getFunction(id)
```
Comment (samarion) [inline core/src/main/scala/stainless/verification/DefaultTactic.scala]: You don't seem to use this.

File: core/src/main/scala/stainless/extraction/imperative/EffectsChecking.scala [hunk 3]
```diff
@@ -12,9 +12,20 @@ trait EffectsChecking { self =>
...
     import effects.symbols._
 
+    def checkPurity(fd: FunAbstraction): Unit = {
+      val funEffects = effects(fd.fullBody)
+
+      if (fd.flags.exists(_.name == "pure") && !funEffects.isEmpty)
+        throw ImperativeEliminationException(fd, s"Function marked @pure cannot have side-effects")
+
+      if (funEffects.exists(_.flags.exists(_.name == "pure")))
+        throw ImperativeEliminationException(fd, s"Functions cannot mutate closed-over variables marked @pure")
```
Comment (samarion) [inline core/src/main/scala/stainless/extraction/imperative/EffectsChecking.scala]: This sounds like a weird use-case.

File: core/src/main/scala/stainless/extraction/imperative/Trees.scala (hunk 1 above)
Comment (romac) [reply core/src/main/scala/stainless/extraction/imperative/Trees.scala]: Yeah I don't know why I added that… :p

File: core/src/main/scala/stainless/verification/DefaultTactic.scala (hunk 2 above)
Comment (romac) [reply core/src/main/scala/stainless/verification/DefaultTactic.scala]: Good catch

File: core/src/main/scala/stainless/extraction/imperative/EffectsChecking.scala (hunk 3 above)
Comment (romac) [reply core/src/main/scala/stainless/extraction/imperative/EffectsChecking.scala]: I don't recall properly anymore, but I think it had something to do with preparing support for inner classes, where we would be annotating variables closed over by a class as `@pure`. I will remove it in the meantime, and rethink this through once I get back to inner classes.

File: core/src/main/scala/stainless/extraction/imperative/ImperativeCleanup.scala [hunk 4]
```diff
@@ -44,12 +53,24 @@ trait ImperativeCleanup extends inox.ast.SymbolTransformer { self =>
               t.Let(r, transform(rhs),
                 recons(l.toVariable, r.toVariable)).copiedFrom(expr)).copiedFrom(expr)
 
+          case s.Variable(id, tpe, flags) =>
+            t.Variable(id, transform(tpe), flags filterNot isImperativeFlag map transform)
+
+          case s.LetRec(defs, body) =>
+            val newDefs = defs map { case s.LocalFunDef(name, tparams, body) =>
+              t.LocalFunDef(transform(name), tparams.map(transform), transform(body).asInstanceOf[t.Lambda])
```
Comment (samarion) [inline core/src/main/scala/stainless/extraction/imperative/ImperativeCleanup.scala]: Isn't this the default implementation of `transform` on `LetRec`?

Comment (romac) [reply core/src/main/scala/stainless/extraction/imperative/ImperativeCleanup.scala]: It should be yeah, I will remove it.

File: core/src/main/scala/stainless/extraction/imperative/ImperativeCleanup.scala [hunk 5]
```diff
@@ -44,12 +53,24 @@ trait ImperativeCleanup extends inox.ast.SymbolTransformer { self =>
...
+            }
+            t.LetRec(newDefs, transform(body))
+
+          // Cleanup leftover `old` nodes if they were used on a reference which is not actually mutated.
+          case s.Old(e) => transform(e)
```
[the remark this answers is not in the document]
Comment (romac) [reply core/src/main/scala/stainless/extraction/imperative/ImperativeCleanup.scala]: Ok, will remove it.

**Pull request merged**